mParticle-Singular/MPKitSingular.m

```diff
@@ -0,0 +1,209 @@
+#define SECRET_KEY @"secret"
+#define DDL_TIMEOUT @"ddlTimeout"
+#define TOTAL_PRODUCT_AMOUNT @"Total Product Amount"
+
+
+
+NSString *appKey;
+NSString *secret;
+int ddlTimeout = 60;
+
+/*
+ mParticle will supply a unique kit code for you. Please contact our team
+ */
++ (NSNumber *)kitCode {
+    return @119;
+}
+
++ (void)load {
+    MPKitRegister *kitRegister = [[MPKitRegister alloc] initWithName:@"Singular" className:@"MPKitSingular" startImmediately:YES];
+    [MParticle registerExtension:kitRegister];
+}
+
+#pragma mark - MPKitInstanceProtocol methods
+
+#pragma mark Kit instance and lifecycle
+- (nonnull instancetype)initWithConfiguration:(nonnull NSDictionary *)configuration startImmediately:(BOOL)startImmediately {
+    self = [super init];
+    
+    if(configuration[API_KEY] != nil)
+        appKey = configuration[API_KEY];
+    if(configuration[SECRET_KEY] != nil)
+        secret = configuration[SECRET_KEY];
+    if(configuration[DDL_TIMEOUT] != nil){
+        ddlTimeout = [configuration[DDL_TIMEOUT] intValue];
+        [Singular setDeferredDeepLinkTimeout:ddlTimeout];
+    }
+    
+    if (!self || !appKey) {
+        return nil;
+    }
+    
+    _configuration = configuration;
+    
+    if (startImmediately) {
+        [self start];
+    }
+    
+    return self;
+}
+
+- (void)start{
+    static dispatch_once_t kitPredicate;
+    dispatch_once(&kitPredicate, ^{
+        /*
+         Start your SDK here. The configuration dictionary can be retrieved from self.configuration
+         */
+        [Singular startSession:appKey withKey:secret];
+        _started = YES;
+        
+        dispatch_async(dispatch_get_main_queue(), ^{
+            NSDictionary *userInfo = @{mParticleKitInstanceKey:[[self class] kitCode]};
+            
+            [[NSNotificationCenter defaultCenter] postNotificationName:mParticleKitDidBecomeActiveNotification
+                                                                object:nil
+                                                              userInfo:userInfo];
+        });
+    });
+}
+
+- (id const)providerKitInstance {
+    if (![self started]) {
+        return nil;
+    }
+    
+    /*
+     If your company SDK instance is available and is applicable (Please return nil if your SDK is based on class methods)
+     */
+    BOOL kitInstanceAvailable = NO;
+    if (kitInstanceAvailable) {
+        /* Return an instance of your company's SDK (if applicable) */
+        return nil;
+    } else {
+        return nil;
+    }
+}
+
+
+#pragma mark Application
+/*
+ Implement this method if your SDK retrieves deep-linking information from a remote server and returns it to the host app
+ */
+- (MPKitExecStatus *)checkForDeferredDeepLinkWithCompletionHandler:(void(^)(NSDictionary *linkInfo, NSError *error))completionHandler {
+    [Singular registerDeferredDeepLinkHandler:^(NSString *deeplink) {
+        NSDictionary *ddlLink = [[NSDictionary alloc] initWithObjectsAndKeys:deeplink,@"deepLink", nil];
```

Comment:
@anismiles rather than using `"deepLink"`, please:

1. Define a public constant that developers can access, exposing it in the header/interface of the kit, for example `SingularDeepLinkKey`.
2. Define it as a namespace string, such as `@"SingularDeepLink"`

That way developers can know if this deep link came from your SDK.